pdns/dnsdistdist/doh.cc

```diff
@@ -0,0 +1,718 @@
+       to store into the IDS, but not for insertion into the
+       rings for example */
+    struct timespec queryRealTime;
+    struct timespec now;
+    gettime(&now);
+    gettime(&queryRealTime, true);
+    char* query = (char*) du->query.c_str();
+
+    struct dnsheader* dh = reinterpret_cast<struct dnsheader*>((char*)query);
+    uint16_t len = du->query.length();
+
+    if (!checkQueryHeaders(dh)) {
+      return -1; // drop
+    }
+
+    string poolname;
+    int delayMsec = 0;
+    const uint16_t * flags = getFlagsFromDNSHeader(dh);
+    const uint16_t origFlags = *flags;
+    uint16_t qtype, qclass;
+    unsigned int consumed = 0;
+    DNSName qname(query, len, sizeof(dnsheader), false, &qtype, &qclass, &consumed);
+    DNSQuestion dq(&qname, qtype, qclass, consumed, &du->dest, &du->remote, dh, 1500, len, false, &queryRealTime);
+    dq.du = du;
+    queryId = ntohs(dh->id);
+    if (!processQuery(holders, dq, poolname, &delayMsec, now))
+    {
+      return -1;
+    }
+
+    if(dq.dh->qr) { // something turned it into a response
+      restoreFlags(dh, origFlags);
+
+      char* response = query;
+      uint16_t responseLen = dq.len;
+      
+      DNSResponse dr(dq.qname, dq.qtype, dq.qclass, consumed, dq.local, dq.remote, reinterpret_cast<dnsheader*>(response), dq.size, responseLen, false, &queryRealTime);
+#ifdef HAVE_PROTOBUF
+      dr.uniqueId = dq.uniqueId;
+#endif
+      dr.qTag = dq.qTag;
+      
+      if (!processResponse(holders.selfAnsweredRespRulactions, dr, &delayMsec)) {
+        return -1;
+      }
+      
+      du->query = std::string(response, responseLen);
+      send(du->rsock, &du, sizeof(du), 0);
+      
+      g_stats.selfAnswered++;
+      doLatencyStats(0);  // we're not going to measure this
+
+      return 0;
+    }
+
+    DownstreamState* ss = nullptr;
+    std::shared_ptr<ServerPool> serverPool = getPool(*holders.pools, poolname);
+    std::shared_ptr<DNSDistPacketCache> packetCache = serverPool->packetCache;
+    auto policy = *(holders.policy);
+    if (serverPool->policy != nullptr) {
+      policy = *(serverPool->policy);
+    }
+    auto servers = serverPool->getServers();
+    if (policy.isLua) {
+      std::lock_guard<std::mutex> lock(g_luamutex);
+      ss = policy.policy(servers, &dq).get();
+    }
+    else {
+      ss = policy.policy(servers, &dq).get();
+    }
+
+    bool ednsAdded = false;
+    bool ecsAdded = false;
+    if (dq.useECS && ((ss && ss->useECS) || (!ss && serverPool->getECS()))) {
+      if (!handleEDNSClientSubnet(dq, &ednsAdded, &ecsAdded, false)) {
+        vinfolog("Dropping query from %s because we couldn't insert the ECS value", du->remote.toStringWithPort());
+        
+        return -1;
+      }
+    }
+
+    uint32_t cacheKey = 0;
+    if (packetCache && !dq.skipCache) {
+
+      uint32_t allowExpired = ss ? 0 : g_staleCacheEntriesTTL;
+      boost::optional<Netmask> subnet;
+      char cquery[1500];
+      memcpy(cquery, query, du->query.length());
```

Comment:
I don't _think_ you need to copy the query here, unless I missing something, since either it will be overwritten by the cached response on a hit, or we won't use `cquery` again before leaving this block.